File: 19. AddWithoutPlus.cpp
// Topic: Add two numbers without using + or - (bitwise)
#include <bits/stdc++.h>
using namespace std;

int addWithoutPlus(int a, int b)
{
    while (b != 0)
    {
        int carry = (a & b) << 1;
        a = a ^ b;
        b = carry;
    }
    return a;
}

int main()
{
    cout << "15 + 27 = " << addWithoutPlus(15, 27) << '\n';
    return 0;
}

// Time Complexity: O(log M) where M is max(a,b)
// Space Complexity: O(1)

File: 19. AddWithoutPlus.py
# Topic: Add two numbers without using + or - (bitwise)

def add_without_plus(a, b):
    mask = 0xFFFFFFFF
    while b != 0:
        carry = (a & b) << 1
        a = (a ^ b) & mask
        b = carry & mask
    return a if a <= 0x7FFFFFFF else ~(a ^ mask)

if __name__ == '__main__':
    print('15 + 27 =', add_without_plus(15,27))

# Time Complexity: O(log M)
# Space Complexity: O(1)
